usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    int rc;
+
+    /* Open current conf file for read */
+    snprintf(src_file, PATH_MAX, "%s/%s", location, "opencryptoki.conf");
+    fp_r = fopen(src_file, "r");
+    if (!fp_r) {
+        TRACE_ERROR("Cannot open %s\n", src_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Open new conf file for write */
+    snprintf(dst_file, PATH_MAX, "%s/%s", location, "opencryptoki.conf_new");
+    fp_w = fopen(dst_file, "w");
+    if (!fp_w) {
+        TRACE_ERROR("Cannot open %s\n", dst_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Insert/replace tokversion parm in new file */
+    snprintf(slot, sizeof(slot), "slot %ld", slot_id);
+    while (fgets(line, sizeof(line), fp_r)) {
+        if (strstr(line, slot)) {
+            fputs(line, fp_w); // write "slot n"
+            /* insert tokversion before '}' */
+            while (fgets(line, sizeof(line), fp_r)) {
+                if (strstr(line, "}")) {
+                    fputs(parm, fp_w);
+                    fputs(line, fp_w);
+                    break;
+                }
+                /* Don't write existing tokversion to new file */
+                if (!strstr(line, "tokversion"))
+                    fputs(line, fp_w);
+            }
+        } else
+            fputs(line, fp_w);
+    }
+
+    fclose(fp_r);
+    fclose(fp_w);
+    fp_r = NULL;
+    fp_w = NULL;
+
+    /* Rename old conf file */
+    snprintf(fname, sizeof(fname), "%s_OLD_TO_BE_DELETED", src_file);
+    rc = rename(src_file, fname);
+    if (rc) {
+        TRACE_ERROR("Cannot rename %s\n", src_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Rename new file */
+    rc = rename(dst_file, src_file);
+    if (rc) {
+        TRACE_ERROR("Cannot rename %s.\n", dst_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+    if (fp_r)
+        fclose(fp_r);
+    if (fp_w)
+        fclose(fp_w);
+
+    return ret;
+
+}
+
+/**
+ * Checks if a backup for the given repository already exists.
+ */
+CK_BBOOL repository_backup_exists(const char *data_store)
+{
+    char fname[PATH_MAX];
+    struct stat statbuf;
+
+    /* Check if data_store_PKCSTOK_MIGRATE_BACKUP exists */
+    memset(fname, 0, PATH_MAX);
+    snprintf(fname, PATH_MAX, "%s_PKCSTOK_MIGRATE_BACKUP", data_store);
+    if (stat(fname, &statbuf) != 0) {
+        return CK_FALSE;
+    }
+
+    if (!S_ISDIR(statbuf.st_mode)) {
+        TRACE_ERROR("%s already exists but is not a directory.\n", fname);
+        return CK_TRUE;
+    }
+
+    return CK_TRUE;
+}
+
+/**
+ * Copy a file given by name from a src folder to a dst folder.
+ */
+void file_copy(char *dst, const char *src, const char *name)
+{
+    char dst_file[PATH_MAX], src_file[PATH_MAX], buf[4096];
+    FILE *fp_r, *fp_w;
+
+    snprintf(dst_file, PATH_MAX, "%s/%s", dst, name);
+    snprintf(src_file, PATH_MAX, "%s/%s", src, name);
+
+    fp_r = fopen(src_file, "r");
```

Comment:
error checking?

Reply:
I guess the user would like to know if the backup really worked or not. Ignoring this error can be very dangerous!